docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,196 @@
+
+- Share common behavior for Machines in MachinePools, which otherwise would need to be implemented by each infrastructure provider.
+- Allow both Cluster API-driven and externally driven interactions with individual Machines in a MachinePool.
+- Enable MachinePools to use deployment strategies similarly to MachineDeployments.
+
+## Motivation
+
+MachineDeployments and Machines provide solutions for common scenarios such as deployment strategy, machine health checks, and retired node deletion. Exposing MachinePool Machines at the CAPI level will enable similar integrations with external projects. 
+
+For example, Cluster Autoscaler needs to delete individual Machines. Without a representation in the CAPI API of individual Machines in MachinePools, projects building integrations with Cluster API would need to be specialized to delete the cloud-specific resources for each MachinePool infrastructure provider. By exposing MachinePool Machines in CAPI, Cluster Autoscaler can treat Machines built by MachineDeployments and MachinePools in a uniform fashion through CAPI APIs.
+
+As another example, currently each MachinePool infrastructure provider must implement their own node drain functionality. When MachinePool Machines become available in the CAPI API, all providers can use its existing node drain implementation.
+
+### Goals
+
+- Enable CAPI MachinePool to create Machines to represent infrastructure provider MachinePool Machines.
+- Extend CAPI MachinePool to include deployment strategies mirroring those of MachineDeployments
+- Ensure that Machines in MachinePools participate in the same behaviors as MachineDeployment Machines (node drain, machine health checks, retired node deletion)
+
+### Non-Goals/Future Work
+
+- Require infrastructure providers to implement MachinePool Machines.
+- Implement Cluster Autoscaler MachinePool integration (a separate task, although we have a working POC)
+
+## Proposal
+
+To create MachinePool Machines, a MachinePool in CAPI needs information about the instances or replicas associated with the provider's implementation of the MachinePool. This information is attached to the MachinePool as a new optional `infrastructureRefList` field, which is populated by the infrastructure provider.
+
+If the `infrastructureRefList` field is populated, the MachinePool will create MachinePool Machines. A MachinePool Machine is implemented as a Cluster API Machine that is owned by a MachinePool. If the field is empty, no MachinePool Machines will be created, and the overall behavior is the same as before this proposal was implemented.
+
+It is the responsibility of each provider to populate the `infrastructureRefList` field and to create provider-specific MachinePool Machine resources behind each Machine. For example, the Docker provider will create a DockerMachinePoolMachine to represent the container instance behind the Machine in the infrastructure provider's MachinePool.
```

Comment:
are there any expected changes to the infrastructure provider contract?

Reply:
Yes, the `InfrastructureMachineSelector` is added to [InfrastructureMachinePool's status](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/machine-pool.html#infrastructure-provider). I'll note that in the proposal and then we can change the provider contract with the Docker reference implementation PR.